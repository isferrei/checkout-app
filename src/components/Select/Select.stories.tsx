import type { Meta, StoryObj } from "@storybook/react";

import { Select } from "./Select";

const meta = {
  title: "Example/Select",
  component: Select,
  tags: ["autodocs"],
  parameters: {
    layout: "fullscreen",
  },
} satisfies Meta;

export default meta;
type Story = StoryObj;

export const Default = {
  args: {
    placeholder: "Selecione um item",
    options: [
      { value: "fox", label: "🦊 Fox" },
      { value: "Butterfly", label: "🦋 Butterfly" },
      { value: "Honeybee", label: "🐝 Honeybee" },
    ],
    onSelect: (selected: any) => console.log("Selected:", selected),
  },
} satisfies Story;
