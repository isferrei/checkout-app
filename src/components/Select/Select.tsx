import React from "react"
import { useState } from "react"
import type {
    ActionMeta,
    GroupBase,
    IndicatorSeparatorProps,
    StylesConfig,
} from "react-select"

import ReactSelect, { components } from "react-select"

export type Option = {
    value: string | number
    label: string | number
}

export type ISelectProps = {
    label: string
    options: Array<{
        label: string | number
        value: string | number
    }>
    placeholder: string
    onSelect: (selected: {
        label: string | number
        value: string | number
    }) => void
}

const indicatorSeparatorStyle = {
    width: 0,
}

const IndicatorSeparator = ({
    innerProps,
}: IndicatorSeparatorProps<unknown, boolean, GroupBase<unknown>>) => {
    return <span style={indicatorSeparatorStyle} {...innerProps} />
}

const CustomSelect = (props: ISelectProps) => {
    const [selected, setSelected] = useState<Option>()

    const handleChange = (
        newValue: Option | null,
        actionMeta: ActionMeta<Option>
    ) => {
        if (newValue === null || newValue === undefined) {
            return
        } else {
            setSelected(newValue)
            props.onSelect(newValue)
        }
    }

    const customStyles: StylesConfig = {
        control: (styles) => ({
            ...styles,
            borderRadius: 0,
            border: "none",
            borderBottom: "1px solid #F4F3F6",
            color: "#666173",
            fontSize: "0.93rem",
        }),
        option: (styles) => ({
            ...styles,
            color: "#666173",
        }),
        singleValue: (styles) => ({
            ...styles,
            color: "#666173",
        }),
        placeholder: (styles) => ({
            ...styles,
            color: "#C9C5D4",
            fontSize: "1rem",
        }),
        container: (styles) => ({
            ...styles,
            width: "100%",
        }),
    }

    return (
        <div>
            <label
                id="select-label"
                className="text-darkGray text-[0.8rem]"
                htmlFor="select"
            >
                {props.label || ""}
            </label>
            <ReactSelect
                id="select"
                name="Select"
                value={selected}
                options={props.options}
                // @ts-ignore
                onChange={(newValue, actionMeta) => setSelected(newValue)}
                placeholder={props.placeholder}
                components={{ IndicatorSeparator }}
                theme={(theme) => ({
                    ...theme,
                    borderRadius: 0,
                })}
                styles={customStyles}
            />
        </div>
    )
}

export function Select(props: ISelectProps) {
    return (
        <CustomSelect
            data-testid="select-element"
            onSelect={props.onSelect}
            options={props.options}
            placeholder={props.placeholder}
            label={props.label}
        />
    )
}
