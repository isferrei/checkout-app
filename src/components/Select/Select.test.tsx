import { cleanup, render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { Select, ISelectProps } from "./Select";
import { act } from "react-dom/test-utils";

afterEach(cleanup);

describe("Select component", () => {
  it("selects an option correctly", () => {
    const options = [
      { value: "option1", label: "Option 1" },
      { value: "option2", label: "Option 2" },
      { value: "option3", label: "Option 3" },
    ];

    const onSelect = jest.fn();

    const selectProps: ISelectProps = {
      label: "Select",
      options,
      placeholder: "Select an option",
      onSelect,
    };

    render(<Select {...selectProps} />);

    const selectElement = screen.getByRole("combobox");
    expect(selectElement).toBeInTheDocument();

    act(() => {
      userEvent.click(selectElement);
    });

    const optionElement = screen.getByText(
      (content, element) =>
        content.startsWith("Option 2") &&
        element?.tagName.toLowerCase() === "div"
    );
    expect(optionElement).toBeInTheDocument();

    act(() => {
      userEvent.click(optionElement);
    });

    const selectedOption = screen.getByText(/Option 2/i);
    expect(selectedOption).toBeInTheDocument();

    expect(selectElement).toMatchSnapshot();
  });
});
