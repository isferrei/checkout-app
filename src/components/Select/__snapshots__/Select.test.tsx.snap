// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select component selects an option correctly 1`] = `
<input
  aria-autocomplete="list"
  aria-expanded="false"
  aria-haspopup="true"
  autocapitalize="none"
  autocomplete="off"
  autocorrect="off"
  class=""
  id="react-select-2-input"
  role="combobox"
  spellcheck="false"
  style="opacity: 0; width: 100%; grid-area: 1 / 2; min-width: 2px; border: 0px; margin: 0px; outline: 0; padding: 0px;"
  tabindex="0"
  type="text"
  value=""
/>
`;
